### PR TITLE
Fix to disable PublishView until ready

### LINKED ISSUE #196: Publish dialog should be disabled while User/Organizations are loading

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fixes #196 

Made use of `BeginDisabledGroup`/`EndDisabledGroup`.

### CODE REVIEW

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 1]
```diff
@@ -63,6 +65,8 @@ private void PopulateView()
                 {
                     Logger.Trace("GetCurrentUser");
 
+                    isLoading = true;
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Why not use `isBusy` for this?

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 2]
```diff
@@ -127,111 +133,111 @@ public override void OnGUI()
...
+                GUILayout.BeginVertical();
                 {
-                    GUI.enabled = !isBusy;
-                    togglePrivate = GUILayout.Toggle(togglePrivate, "Create as a private repository");
-                    GUI.enabled = true;
-                }
-                GUILayout.EndHorizontal();
+                    GUILayout.BeginHorizontal();
+                    {
+                        togglePrivate = GUILayout.Toggle(togglePrivate, "Create as a private repository");
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Move this string to a const while you're at it

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 3]
```diff
@@ -127,111 +133,111 @@ public override void OnGUI()
...
-            }
-            GUILayout.EndHorizontal();
+                GUILayout.EndHorizontal();
 
-            GUILayout.Label("Description");
-            GUI.enabled = !isBusy;
-            repoDescription = EditorGUILayout.TextField(repoDescription);
-            GUI.enabled = true;
-            GUILayout.Space(Styles.PublishViewSpacingHeight);
+                GUILayout.Label("Description");
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Move this string to a const while you're at it

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 4]
```diff
@@ -127,111 +133,111 @@ public override void OnGUI()
...
 
-                GUILayout.BeginVertical();
-                {
-                    GUILayout.Label("Repository Name");
-                    GUI.enabled = !isBusy;
-                    repoName = EditorGUILayout.TextField(repoName);
-                    GUI.enabled = true;
+                    GUILayout.BeginVertical();
+                    {
+                        GUILayout.Label("Repository Name");
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Move this string to a const while you're at it

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 5]
```diff
@@ -127,111 +133,111 @@ public override void OnGUI()
...
-            GUILayout.BeginHorizontal();
+            EditorGUI.BeginDisabledGroup(isLoading || isBusy);
             {
-                GUILayout.BeginVertical();
+                GUILayout.BeginHorizontal();
                 {
-                    GUILayout.Label("Owner");
+                    GUILayout.BeginVertical();
+                    {
+                        GUILayout.Label("Owner");
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Move this string to a const while you're at it

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs (hunk 1 above)
Comment (StanleyGoldman) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: I don't have a good reason for using a different flag, mental clarity if nothing else...

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 6]
```diff
@@ -91,6 +98,8 @@ private void PopulateView()
 
                             owners = owners.Union(organizationLogins).ToArray();
                         });
+                    }).Finally(task => {
+                        isBusy = false;
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Oh, this needs to be `FinallyInUI`, serialized ui data should only be touched in thread

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs [hunk 7]
```diff
@@ -127,111 +136,111 @@ public override void OnGUI()
...
+                            }
 
-                        GitClient.RemoteAdd("origin", repository.CloneUrl)
-                                 .Then(GitClient.Push("origin", Repository.CurrentBranch.Value.Name))
-                                 .ThenInUI(Parent.Finish)
-                                 .Start();
-                    }, organization);
+                            GitClient.RemoteAdd("origin", repository.CloneUrl)
+                                     .Then(GitClient.Push("origin", Repository.CurrentBranch.Value.Name))
+                                     .ThenInUI(Parent.Finish)
```
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Before doing this, `isBusy` should be reset to false. We shouldn't be calling `Parent.Finish` here anyway, we should be calling `Finish` (which calls `Parent.Finish`), and then we can reset the busy flag there.

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs (hunk 6 above)
Comment (StanleyGoldman) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: Hmm, the more I look at it, I realize this `Finally` is incorrect in other ways..
The `GetCurrentUser` callback has two different execution paths, one is synchronous the other is async.

https://github.com/github-for-unity/Unity/blob/0e489246486557981c314f7c0c82a65bc65fe1b3/src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PublishView.cs#L73-L103

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs (hunk 7 above)
Comment (StanleyGoldman) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: > Before doing this, isBusy should be reset to false.

😕 I dont understand what you mean by this

> We shouldn't be calling Parent.Finish here anyway, we should be calling Finish (which calls Parent.Finish), and then we can reset the busy flag there.

👍 But the rest of this I got

File: src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs (hunk 6 above)
Comment (shana) [inline src/UnityExtension/Assets/Editor/GitHub.Unity/UI/PUblishView.cs]: aaaah yes we're not in the new task system here. bummer. Ok we can't use `Finally` here, it's on the wrong thread. Let's set `isBusy = false` in all the return points inside the callbacks for now and then talk about how this API should actually look like separately.

File: src/GitHub.Api/Application/ApiClient.cs [hunk 8]
```diff
@@ -174,16 +182,15 @@ public ApiClient(UriString hostUrl, IKeychain keychain, IGitHubClient githubClie
...
-        private async Task CreateRepositoryInternal(NewRepository newRepository, Action<Octokit.Repository, Exception> callback, string organization)
+        private async Task<Octokit.Repository> CreateRepositoryInternal(NewRepository newRepository, string organization)
         {
             try
             {
                 logger.Trace("Creating repository");
 
                 if (!await EnsureKeychainLoaded())
                 {
-                    callback(null, new Exception("Keychain Not Loaded"));
```
Comment (shana) [inline src/GitHub.Api/Application/ApiClient.cs]: You need to throw here to have the same behaviour as before

Comment (StanleyGoldman) [inline src/GitHub.Api/Application/ApiClient.cs]: I mean, the user shouldn't reach here without having their Keychain loaded.
So i guess this code being here is more of a formality...

It's something I hope to clean up as we..
1. Get proper tasks around the API
2. Figure out a clean way to handle authentication errors from the API and Git

Comment (shana) [inline src/GitHub.Api/Application/ApiClient.cs]: Right, yeah, kinda need to rewrite this whole thing...

**Pull request merged**